crates/turbopack-css/src/chunk/writer.rs

```diff
@@ -1,108 +1,156 @@
+
+                let content = chunk_item.content().await?;
+                code.push_source(&content.inner_code, content.source_map, &indenter)
+                    .await?;
+                indenter.pop(*indent);
+                code.push_str(&format!("\n{}\n", close), &indenter)?;
+
+                result.codes.push(code.build().cell());
                 stack.pop();
             }
         }
     }
 
-    Ok(external_imports)
+    Ok(result.cell())
 }
 
-pub struct WriterWithIndent<T: Write> {
-    writer: T,
-    indent_str: String,
+struct CodeBuilderWithIndent {
+    code: CodeBuilder,
     needs_indent: bool,
 }
 
-impl<T: Write> WriterWithIndent<T> {
-    pub fn new(buffer: T) -> Self {
+impl Default for CodeBuilderWithIndent {
+    fn default() -> Self {
         Self {
-            writer: buffer,
-            indent_str: "".to_string(),
+            code: CodeBuilder::default(),
             needs_indent: true,
         }
     }
+}
 
-    pub fn push_indent(&mut self, indent: usize) -> std::fmt::Result {
-        self.indent_str += &" ".repeat(indent);
+#[derive(Default)]
+struct Indent(String);
 
-        Ok(())
+impl Indent {
+    pub fn push(&mut self, n: usize) {
+        self.0 += &" ".repeat(n);
     }
 
-    pub fn pop_indent(&mut self, indent: usize) -> std::fmt::Result {
-        self.indent_str = " ".repeat(self.indent_str.len().saturating_sub(indent));
+    pub fn pop(&mut self, n: usize) {
+        self.0 = " ".repeat(self.0.len().saturating_sub(n));
+    }
 
-        Ok(())
+    pub fn to_str(&self) -> &str {
+        &self.0
     }
 }
 
-impl<T: Write> Write for WriterWithIndent<T> {
-    #[inline]
-    fn write_str(&mut self, s: &str) -> std::fmt::Result {
-        for c in s.chars() {
-            self.write_char(c)?;
-        }
+impl CodeBuilderWithIndent {
+    pub async fn push_source(
+        &mut self,
+        code: &Rope,
+        map: Option<ParseResultSourceMapVc>,
+        indent: &Indent,
+    ) -> Result<()> {
+        let indented = self.wrap_indent(&code.to_str()?, indent)?;
+        let indented_map = if let Some(map) = map {
+            Some(
+                ParseResultSourceMap::with_indent(&*map.await?, indent.to_str().len() as u32)
+                    .cell()
+                    .as_generate_source_map(),
+            )
+        } else {
+            None
+        };
+        self.code.push_source(&indented, indented_map);
+        Ok(())
+    }
 
+    pub fn push_str(&mut self, code: &str, indent: &Indent) -> Result<()> {
+        let indented = self.wrap_indent(code, indent)?;
+        self.code.push_source(&indented, None);
         Ok(())
     }
 
-    #[inline]
-    fn write_char(&mut self, c: char) -> std::fmt::Result {
-        if c == '\n' {
-            self.writer.write_char('\n')?;
-            self.needs_indent = true;
+    fn wrap_indent(&mut self, code: &str, indent: &Indent) -> Result<Rope> {
+        let mut indented = RopeBuilder::default();
+        for c in code.chars() {
```

Comment:
I think this can be done as a `for line in code.split('\n')`, which'll allow us to write large chunks of chars at a single time.